compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -0,0 +1,402 @@
+                val thisVar = irTemporaryVarDeclaration(
+                    worker.returnType, nameHint = "\$this", isMutable = false
+                )
+                addMapping(constructor.constructedClass.thisReceiver!!.symbol, thisVar)
+
+                constructor.body?.contents?.forEach { statement ->
+                    +statement
+                        .transform(object : IrElementTransformerVoid() {
+                            // Don't recurse under nested class declarations
+                            override fun visitClass(declaration: IrClass): IrStatement {
+                                return declaration
+                            }
+
+                            // Capture the result of a delegating constructor call in a temporary variable "thisVar".
+                            //
+                            // Within the constructor we replace references to "this" with references to "thisVar".
+                            // This is safe, since the delegating constructor call precedes all references to "this".
+                            override fun visitDelegatingConstructorCall(expression: IrDelegatingConstructorCall): IrExpression {
+                                expression.transformChildrenVoid()
+                                return irSetVar(thisVar.symbol, expression)
+                            }
+
+                            // A constructor body has type unit and may contain explicit return statements.
+                            // These early returns may have side-effects however, so we still have to evaluate
+                            // the return expression. Afterwards we return "thisVar".
+                            // For example, the following is a valid inline class declaration.
+                            //
+                            //     inline class Foo(val x: String) {
+                            //       constructor(y: Int) : this(y.toString()) {
+                            //         if (y == 0) return throw java.lang.IllegalArgumentException()
+                            //         if (y == 1) return
+                            //         return Unit
+                            //       }
+                            //     }
+                            override fun visitReturn(expression: IrReturn): IrExpression {
+                                expression.transformChildrenVoid()
+                                if (expression.returnTargetSymbol != constructor.symbol)
+                                    return expression
+
+                                return irReturn(irBlock(expression.startOffset, expression.endOffset) {
+                                    +expression.value
+                                    +irGet(thisVar)
+                                })
+                            }
+                        }, null)
+                        .transform(this@JvmInlineClassLowering, null)
+                        .patchDeclarationParents(worker)
```

Comment:
Why is `patchDeclarationParents` needed here?

Reply:
Because we copy the statement from the constructor to a static replacement function. It's possible to do this inline in one of the other traversals, but this just ends up duplicating code which is already in `patchDeclarationParents`.